compiler/rustc_metadata/src/locator.rs

```diff
@@ -499,18 +511,39 @@ impl<'a> CrateLocator<'a> {
         rlibs: FxIndexMap<PathBuf, PathKind>,
         rmetas: FxIndexMap<PathBuf, PathKind>,
         dylibs: FxIndexMap<PathBuf, PathKind>,
+        interfaces: FxIndexMap<PathBuf, PathKind>,
     ) -> Result<Option<(Svh, Library)>, CrateError> {
         let mut slot = None;
         // Order here matters, rmeta should come first.
         //
         // Make sure there's at most one rlib and at most one dylib.
         //
         // See comment in `extract_one` below.
-        let source = CrateSource {
-            rmeta: self.extract_one(rmetas, CrateFlavor::Rmeta, &mut slot)?,
-            rlib: self.extract_one(rlibs, CrateFlavor::Rlib, &mut slot)?,
-            dylib: self.extract_one(dylibs, CrateFlavor::Dylib, &mut slot)?,
+        let rmeta = self.extract_one(rmetas, CrateFlavor::Rmeta, &mut slot)?;
+        let rlib = self.extract_one(rlibs, CrateFlavor::Rlib, &mut slot)?;
+
+        // We assume here that the interface and the dylib are located
+        // in the same directory.
+        let interface = self.extract_one(interfaces, CrateFlavor::SDylib, &mut slot)?;
+        let sdylib = if let Some(interface) = interface {
+            let suffix = self
+                .target
+                .dll_suffix
+                .as_ref()
+                .strip_prefix(".")
+                .unwrap_or(self.target.dll_suffix.as_ref());
+            let dylib = interface.0.with_extension(suffix);
+            if !dylib.exists() {
+                return Err(CrateError::ExternLocationNotExist(self.crate_name, dylib));
```

Comment:
Ideally, all these errors should be covered by tests (unless they occur due to some external system failures that are hard to reproduce).